File: LeetCode/2966/golang/lc2966.go
package lc2966

import "slices"

// Runtime complexity: O(sort) + O(n)
// Auxiliary space: O(sort)
func divideArray(nums []int, k int) [][]int {
	SZ := len(nums)
	slices.Sort(nums)
	ans := make([][]int, 0, SZ/3)
	for i := 0; i < SZ; i += 3 {
		if nums[i+2]-nums[i] > k {
			// Can not divide the numbers into this subarray; so the partition can not be done at all.
			// Return an empty slice.
			return [][]int{}
		}
		ans = append(ans, []int{nums[i], nums[i+1], nums[i+2]})
	}
	return ans
}

File: LeetCode/2966/cpp/lc2966.cpp
#include <algorithm>
#include <cassert>
#include <vector>

class Solution {
public:
  std::vector<std::vector<int>> divideArray(std::vector<int>& nums, int k) {
    std::ranges::sort(nums);
    std::vector<std::vector<int>> ans;
    ans.reserve(nums.size() / 3); // sunny day scenario. Will not use all that memory if problem is unsolvable.
    for (int i = 0; i < nums.size() - 2; i += 3) {
      if (nums[i + 2] - nums[i] > k) {
        // Can not divide the numbers into this subarray => return empty array.
        return {};
      }
      ans.push_back({nums[i], nums[i + 1], nums[i + 2]});
    }
    return ans;
  }
};

int main() {
  Solution s;
  {
    std::vector<int> nums = {1, 3, 4, 8, 7, 9, 3, 5, 1};
    std::vector<std::vector<int>> expected = {{1, 1, 3}, {3, 4, 5}, {7, 8, 9}};
    assert(s.divideArray(nums, 2) == expected);
  }
  {
    std::vector<int> nums = {1, 3, 3, 2, 7, 3};
    std::vector<std::vector<int>> expected = {};
    assert(s.divideArray(nums, 3) == expected);
  }
}
